### PR TITLE
Adding configuration support

### LINKED ISSUE #825: Orchard environmental settings discussion
Gathering some thoughts on how Orchard2 deals with environmental settings in conjunction with a multi-environment release flow.

### Background
We work with the notion "_code flows up, content flows down_", meaning that code changes flow from `Development Environment -> Test Environment -> Production Environment`, whereas CMS content flows in reverse `Production Environment -> Test and Development Environments`.

The reason for this is to be able to code and test against real-world data, of course with the relevant obfuscation of sensitive data stored in the db, including user account data.

### In a Perfect World
All environment-related settings, assets and configuration are in App_Data (e.g. Settings.txt), Web.config or, more recently, appsettings.json. Only "portable", environment-agnostic content is stored in "the database". 

This means that no data transformations are required to restore from Production to Test or Development.

### In the Real World
Environment-related settings do find their way into the db, especially where settings relate to hosts. In Orchard1 a good example is the SecureSocketsLayer module (Secure Host and Insecure Host) and site settings such as UseCdn. In Orchard2 a good example is the OpenId module (Authority, App Redirect URL's, etc.).

This means that flowing the CMS content from Production to downlevel environments is somewhat more complex as these settings need to be updated to match each environment.

### Discussion Points
1. Is this an issue worth discussion and, if so:
2. Is Settings.txt being under-utilised in favour of db-stored settings?
3. Should the guideline be to _not_ store any environment-related settings in the db?
4. What would be the definition of an environment-related setting?
5. Is there an overall better solution to this problem?

### LINKED ISSUE #1456: Reimplement FileShellSettingsConfigurationProvider
Currently, `FileShellSettingsConfigurationProvider` uses ASP.NET Core's Configuration API to load shell settings. However, it relies on a configuration provider that reads from **tenants.json**. This is not the right implementation, since the Configuration API is much more general purpose, and we could just as well define tenants in **appsettings.json** (no need to create a configuration provider that reads from tenants.json, which has a specific scope in terms of the kinds of settings expected there).

Better would be to update `FileShellSettingsConfigurationProvider` such that it parses "tenants.json" directly (not via the Configuration API), and perhaps implement another shell settings provider that does use the Configuration API.

See also discussion in #1451.

### LINKED ISSUE #2038: Provide a per-tenant configuration story
See http://www.paraesthesia.com/archive/2018/06/20/microsoft-extensions-configuration-deep-dive/ for more details on how to share configuration for all tenants, and customize per tenant. And also support ENV.

There must be another issue I created for that but can't find it.

### LINKED ISSUE #2051: Use json for site settings
And remove YAML completely as their should not be any reason to use it after this change.

### PR DESCRIPTION
**Pull request by sebastienros**
Fixes #2038
Fixes #1456
Fixes #825
Fixes #2051

### CODE REVIEW

**Comment by jtkech** [comment]
So, because now tenant names are only resolved through the shell container directories, `tenants.json` renamed to `appsettings.tenants[.EnvironmentName].json` doesn't work.

The 1st idea is to also lookup to these config files to resolve tenants. I will work on it.

**Comment by sebastienros** [comment]
Please don't change anything you are mixing things up. I can explain on Monday.

**Comment by jtkech** [comment]
Okay cool, i will not change anything else, just doing tests to better understand things.

I'm playing with children, sections, how to specify a setting to a specific tenant or for all. In our context, for json files we need to specify the full path, even it is at the root.

For infos, with `tenants.json`, each tenant was using a `ShellSettingsWithTenants` holding a list of features grabbed from the old `settings.Configuration`.

**Comment by rserj** [comment]
I think this is a really useful feature, any updates on this?

**Comment by sebastienros** [comment]
We are actively working on it. @jtkech has a PR for my PR, where he's testing many ways to do it, and my crazy requirements. We are at a design right now that will also allow for faster cold-start when using many tenants.

**Comment by sebastienros** [comment]
I did a demo during yesterday's meeting. The video should be available soon.

**Comment by jtkech** [comment]
Todo: Filter invalid tenant section without the "State" marker field.

File: src/OrchardCore.Build/Dependencies.props [hunk 1]
```diff
@@ -40,6 +40,7 @@
     <PackageManagement Include="xunit.analyzers" Version="0.10.0" />
     <PackageManagement Include="xunit" Version="2.4.0" />
     <PackageManagement Include="xunit.runner.visualstudio" Version="2.4.0" />
+    <PackageManagement Include="YamlDotNet" Version="5.3.0" />
```
Comment (sebastienros) [inline src/OrchardCore.Build/Dependencies.props]: Can you remove this library altogether, and parse the yaml manually. It should just be about reading line by line, and splitting on the name.

File: src/OrchardCore/OrchardCore/OrchardCore.csproj [hunk 2]
```diff
@@ -23,6 +23,7 @@
     <PackageReference Include="Microsoft.Extensions.Localization" />
     <PackageReference Include="Microsoft.Extensions.Logging" />
     <PackageReference Include="NodaTime" />
+    <PackageReference Include="YamlDotNet" />
```
Comment (sebastienros) [inline src/OrchardCore/OrchardCore/OrchardCore.csproj]: Same here

File: src/OrchardCore.Build/Dependencies.props (hunk 1 above)
Comment (jtkech) [reply src/OrchardCore.Build/Dependencies.props]: Okay

File: src/OrchardCore/OrchardCore/OrchardCore.csproj (hunk 2 above)
Comment (jtkech) [reply src/OrchardCore/OrchardCore/OrchardCore.csproj]: Okay

**Comment by sebastienros** [comment]
Can you also merge dev, there are conflict since Antoine improved the Tenants ui

**Comment by jtkech** [comment]
Okay no problem i will do it. Why did he do that just now ;)

I will also comment my last changes and maybe some last choices i'm not sure.

**Comment by jtkech** [comment]
First comments on my last changes.

- So, we use the app config on which we 1st add by default `App_Data/appsettings.{env}.json`. Now through the DI an app level service can override the config sources added at this point.

- Then we reapply at the end env variables and cmd line arguments to build the **base config**.

- Then we use a **"local" / tenant specific** config source, by default `{tenantFolder}/appsettings.json` which is lazily built. Here also an app level service can override the tenant specific config sources.

- When saving a tenant, config values (different from the base config) are saved by default in `{Sites}/{tenant}/appsettings.json`. Now we 1st read the file so that we don't overwrite other configs which may have been added manually. And because of the above, another source can be updated.

- When saving a tenant, we do the same for data which are immediately required to start a tenant, url host and prefix and tenant state, but we save them by default in a shared `App_Data/tenants.json` so that we can use only one source when loading settings. Now we can also override this source.

**Todo**: Remove the dependency on yalm.

**Comment by sebastienros** [comment]
Don't worry about the yaml dependency. We can just file a bug for RC with p0 so we don't forget
I assume that's fine as we actually did remove the yaml files themselves

**Comment by jtkech** [comment]
Okay cool.

Yes, i just kept the reference to use the Yaml Deserializer.

File: src/OrchardCore.Modules/OrchardCore.Media.Azure/Startup.cs [hunk 3]
```diff
@@ -15,9 +15,9 @@ namespace OrchardCore.Media.Azure
     public class Startup : StartupBase
     {
         private ILogger<Startup> _logger;
-        private readonly IShellConfiguration _configuration;
+        private readonly ITenantConfiguration _configuration;
```
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.Media.Azure/Startup.cs]: If you start renaming Shells into Tenants, you will get @PinpointTownes angry

Comment (jtkech) [reply src/OrchardCore.Modules/OrchardCore.Media.Azure/Startup.cs]: Ah okay, no problem i will revert it back to `IShellConfiguration`.

Comment (PinpointTownes) [reply src/OrchardCore.Modules/OrchardCore.Media.Azure/Startup.cs]: @sebastienros as long as it's consistent, I have no problem with that being called "shell" or "tenant" :smile:

Comment (jtkech) [reply src/OrchardCore.Modules/OrchardCore.Media.Azure/Startup.cs]: Too late, already reverted 😄

Comment (PinpointTownes) [reply src/OrchardCore.Modules/OrchardCore.Media.Azure/Startup.cs]: While you're here, would that be worth renaming `TenantState` to `ShellState`? :trollface:

Comment (jtkech) [reply src/OrchardCore.Modules/OrchardCore.Media.Azure/Startup.cs]: Just for infos `ShellState` is an existing class ;)

    public class ShellState
    {
        public List<ShellFeatureState> Features { get; } = new List<ShellFeatureState>();
    }

More seriously @sebastienros @PinpointTownes for naming it is as you want, just let me know.

Comment (sebastienros) [reply src/OrchardCore.Modules/OrchardCore.Media.Azure/Startup.cs]: I would need to resurrect an old chat with Kevin where he explained me the discrepancies and how to fix it. In the meantime no changes.

**Comment by jtkech** [comment]
@sebastienros 

- So, in my last commit, just to show, can be reverted, the suggestion is to lazily create each shell on a 1st request, as done for the pipeline, so that its specific config is effectively lazily built.

    Tried in debug mode with 20 tenants, the very 1st request is much more faster, then, on the 1st request of a given tenant, it adds an average time e.g of 80 ms depending on the recipe used.

- Doing so, listing all shells, as in the tenants index page, may need more time. But, in fact the tenants page only use `ShellSettings` so i updated it to use the recent `_shellHost.GetAllSettings()`. I also fixed a little issue where the bulk action to enable multiple tenants was not working.

- The openId module also uses `ListShellContextsAsync()` in some place to create a scope on each one to see if it is a `Server`, so nothing that we can do here.

---

**For infos**

- Another compromise would be to create shells with full parallelism. We can do it because to 1st get the `ShellDescriptor` we now use a regular scope without swapping `RequestServices`.

    Tried in debug mode with 20 tenants, always a base time of 500 ms, maybe for the 1st shell and / or to resolve app level services. Then an average time e.g of 35 ms per tenant creation in place of 70 ms.

- But in this case, lazily build specific configs would not be so useful and maybe we could get rid of these "local" configs by putting, after a tenant edit / setput, more values in the shared `tenants.json`.

**Comment by PinpointTownes** [comment]
> The openId module also uses ListShellContextsAsync() in some place to create a scope on each one to see if it is a Server, so nothing that we can do here.

If it's too expensive, we may remove that and list all the tenants instead of just the ones that have the server feature enabled. If we don't want to degrade the user experience too much, we may want to make a dynamic check to verify the server feature is enabled once the admin selected a tenant in the available tenants list.

**Comment by jtkech** [comment]
@Jetski5822 @PinpointTownes 

There are both implicitly implemented because we expose an `IShellConfiguration` which inherits of `IConfiguration`. That said the related source needs to be reloadable.

We use the application sources on which we add other sources as e.g a json file under App_Data for all tenants, specific ones under tenant folders ...

Just tried `IOptionsMonitor` by defining an class, using `services.Configure<MyOptions>(_shellConfiguration)` and defining some properties in the application `appsettings.json`, this under the `OrchardCore` section for all tenants and / or under a subsection for a given tenant.

Through an injected `IOptionsMonitor<MyOptions>` in a tenant singleton i was able to update 
the json file and see the change through `options.CurrentValue`. It works because i updated the application json file which by default is reloadable.

It would not work by updating one of our additional sources which are not reloadable by default. But you can override source(s) through app level service(s) which will participate to build the config. E.g you can easily define that the `App_Data/appsettings.json` source will use `reloadOnChange = true`.

**Comment by Jetski5822** [comment]
Heads up - I made a change to the GraphQL stuff and introduced some configuration in to there. So we need to make that change in the PR too

**Comment by PinpointTownes** [comment]
> the json file and see the change through options.CurrentValue. It works because i updated the application json file which by default is reloadable.

:+1:

File: src/OrchardCore/OrchardCore.Abstractions/Shell/Configuration/ITenantConfigurationSources.cs [hunk 4]
```diff
@@ -0,0 +1,20 @@
+using System.Collections.Generic;
+using Microsoft.Extensions.Configuration;
+
+namespace OrchardCore.Environment.Shell.Configuration
+{
+    public interface ITenantConfigurationSources
```
Comment (PinpointTownes) [inline src/OrchardCore/OrchardCore.Abstractions/Shell/Configuration/ITenantConfigurationSources.cs]: Should these interfaces/classes be named `Shell*`? :sweat_smile:

**Comment by jtkech** [comment]
@Jetski5822 Okay, it's noted

**Comment by jtkech** [comment]
@PinpointTownes 

We opted to keep the lazy creating of shells, and now only the background service uses `ListShellContextsAsync()` which now doesn't create a released shell (released or not yet created).

I already updated your driver code to use instead `_shellHost.GetAllSettings()` but right now we still call on all settings `_shellHost.GetScopeAsync(shellSettings)` so all the shells are still created in one pass.

So, would be good to make, as you said, a dynamic check to verify the server feature.

Can be done in another PR as we will merge this one very soon.

**Pull request merged**